xdsl/dialects/csl.py

```diff
@@ -883,6 +925,527 @@ def verify_(self) -> None:
+                    (DsdType, DsdType, DsdType),
+                    (DsdType, Float16Type, DsdType),
+                    (DsdType, DsdType, Float16Type),
+                    (f16_pointer, Float16Type, DsdType),
+                ]
+            case BuiltinOpPrecisionKind.mixed:
+                return []
+            case BuiltinOpPrecisionKind.single:
+                return [
+                    (DsdType, DsdType, DsdType),
+                    (DsdType, Float32Type, DsdType),
+                    (DsdType, DsdType, Float32Type),
+                    (f32_pointer, Float32Type, DsdType),
+                ]
+
+
+@irdl_op_definition
+class FnegOp(_MultiprecisionDsdOp):
+    """
+    Implements @fnegh and @fnegs
+    """
+
+    name = "csl.fneg"
+
+    def get_signatures(self) -> FunctionSignatures:
+        match self.precision.data:
+            case BuiltinOpPrecisionKind.half:
+                return [
+                    (DsdType, DsdType),
+                    (DsdType, Float16Type),
+                ]
+            case BuiltinOpPrecisionKind.mixed:
+                return []
+            case BuiltinOpPrecisionKind.single:
+                return [
+                    (DsdType, DsdType),
+                    (DsdType, Float32Type),
+                ]
+
+
+@irdl_op_definition
+class FnormOp(_MultiprecisionDsdOp):
+    """
+    Implements @fnormh and @fnorms
+    """
+
+    name = "csl.fnorm"
+
+    def get_signatures(self) -> FunctionSignatures:
+        match self.precision.data:
+            case BuiltinOpPrecisionKind.half:
+                return [
+                    (f16_pointer, Float16Type),
+                ]
+            case BuiltinOpPrecisionKind.mixed:
+                return []
+            case BuiltinOpPrecisionKind.single:
+                return [
+                    (f32_pointer, Float32Type),
+                ]
+
+
+@irdl_op_definition
+class Fs2hOp(_BuiltinDsdOp):
+    """
+    Implements @fs2h
+    """
+
+    name = "csl.fs2h"
+
+    def get_signatures(self) -> FunctionSignatures:
+        return [
+            (DsdType, DsdType),
+            (DsdType, Float32Type),
+        ]
+
+
+@irdl_op_definition
+class Fs2xp16Op(_BuiltinDsdOp):
+    """
+    Implements @fs2xp16
+    """
+
+    name = "csl.fs2xp16"
+
+    def get_signatures(self) -> FunctionSignatures:
+        return [
+            (DsdType, DsdType),
+            (DsdType, Float32Type),
+            (i16_pointer, Float32Type),  # todo this might be an error in the CSL docs
+        ]
+
+
+@irdl_op_definition
+class FscaleOp(_MultiprecisionDsdOp):
+    """
+    Implements @fscaleh and @fscales
+    """
+
+    name = "csl.fscale"
+
+    def get_signatures(self) -> FunctionSignatures:
+        match self.precision.data:
+            case BuiltinOpPrecisionKind.half:
+                return [
+                    (f16_pointer, Float16Type, i16_value),
+                ]
+            case BuiltinOpPrecisionKind.mixed:
+                return []
```

Comment:
This is a way of saying "I support no mixed-precision modes", right?

Reply:
yes

Reply:
It will cause it to fail the verification performed by the superclass